### PR TITLE
Add implementation of atomics on CPU using __atomic

### LINKED ISSUE #172: atomics on CPU
I have the following simple atomic counter example that compiles and runs fine for CUDA, but doesn't compile on the CPU.
```
#include <iostream>
#include <CL/sycl.hpp>

namespace s = cl::sycl;

int main()
{
    s::queue q;

    int counter = 0;
    {
        s::buffer<int> counter_buf(&counter, 1);

        q.submit([&](cl::sycl::handler& cgh)
        {
            auto access_counter = counter_buf.get_access<cl::sycl::access::mode::read_write>(cgh);

            cgh.parallel_for<class atomic_increment>(s::range<1>(1 << 30), [=] (cl::sycl::id<1> tid)
            {
                s::atomic<int> atomic_counter { s::global_ptr<int> {&access_counter[0]} };
                atomic_counter.fetch_add(1);
            });
        });
    }

    std::cout << "Counter: " << counter << std::endl;
}
```

```
/opt/hipSYCL/CUDA/bin/syclcc-clang --hipsycl-gpu-arch=sm_52 -O3 atomic-counter.cpp -o atomic-counter-cuda
```
compiles and runs fine.

```
/opt/hipSYCL/CUDA/bin/syclcc-clang --hipsycl-platform=cpu -g atomic-counter.cpp -o atomic-counter-cpu
In file included from atomic-counter.cpp:3:
In file included from /opt/hipSYCL/CUDA/bin/../include/CL/sycl.hpp:58:
/opt/hipSYCL/CUDA/bin/../include/CL/sycl/atomic.hpp:103:12: error: use of undeclared identifier 'atomicAdd'
    return atomicAdd(_ptr, operand);
           ^
atomic-counter.cpp:23:32: note: in instantiation of function template specialization 'cl::sycl::atomic<int, cl::sycl::access::address_space::global_space>::fetch_add<int, nullptr>' requested here
                atomic_counter.fetch_add(1);
```

I know atomics are not fully supported (they are listed as a limitation in README), but other issues suggest that only minor features of atomics are missing. Is the above error by design, or is it something that can be easily fixed?

### PR DESCRIPTION
**Pull request by mrzv**
This is implementation of atomics on CPU using __atomic built-ins. For now, it's WIP for discussion. It's meant to fix #172.

### CODE REVIEW

**Comment by mrzv** [comment]
@jeffhammond What's the right way to implement `fetch_min` and `fetch_max`? There is not an equivalent `__atomic`. The only way I can think of is via `compare_exchange`. Is there a better way?

**Comment by mrzv** [comment]
@jeffhammond Also the memory order on success and failure for compare_exchange_strong deserves attention. I just picked relaxed without knowing what I'm doing.

**Comment by mrzv** [comment]
This PR includes implementation listed in #174.

File: include/CL/sycl/atomic.hpp [hunk 1]
```diff
@@ -81,16 +81,24 @@ class atomic {
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
     return atomicExch(_ptr, operand);
 #else
-    return detail::invalid_host_call_dummy_return<T>();
+    return __atomic_exchange_n(_ptr, operand, __ATOMIC_RELAXED);
```
Comment (illuhad) [inline include/CL/sycl/atomic.hpp]: I think the `#ifdef` code paths should be structured slightly differently. There are currently several compilation paths when compiling with hipSYCL:

* GPU compilation pass
* CPU compilation pass of the CUDA/ROCm backends. This is where we currently have those `invalid_host_call_dummy_return()` calls. Because using CPU and GPU simultaneously is not yet supported (due to limitations in the runtime which will be fixed by the new runtime), these dummy calls should never be actually executed and are in principle something like a `assert(false)`.
* CPU compilation pass when compiling for the CPU backend.

Now, the `__HIPSYCL_DEVICE_CALLABLE__` macro is defined
* by CUDA/ROCm backend if currently in the GPU compilation pass
* by the CPU backend because hipCPU defines many of those HIP backends.
In other words, it's defined when HIP intrinsics like atomicExch are available.

This means that the code here will never execute in practice because it will be compiled for the host compilation pass for the CUDA/ROCm backend.

Instead, we want to have those atomic functions run whenever we compile for CPU. Solution: Check with the `SYCL_DEVICE_ONLY` macro if we are in a GPU compilation pass:
```
#ifdef SYCL_DEVICE_ONLY
  return atomicExch(...);
#else
  return __atomic_exchange_n(...)
#endif
```

**Comment by mrzv** [comment]
One more thing that's unclear: `load` and `store` are not implemented. There are obvious `__atomic` functions to use, but I don't see what the right function is in HIP. Any ideas?

File: include/CL/sycl/atomic.hpp (hunk 1 above)
Comment (mrzv) [reply include/CL/sycl/atomic.hpp]: How about now?

Comment (illuhad) [reply include/CL/sycl/atomic.hpp]: Fine now, we could even drop the surrounding `#ifdef __HIPSYCL_DEVICE_CALLABLE__` to make the code simpler. There are enough other places that would prevent running CPU code in kernels when compiling for GPU. In the future I want to make the CPU backend always available (even when compiling for GPU), so then we would have to remove this anyway.

**Comment by illuhad** [comment]
@mrzv I think you may have discovered the reason why those functions were unimplemented ;) I would suggest to only implement those on CPU for now.
I will investigate how to implement those functions on GPU. If I remember correctly, the spec was very unclear for those anyway because it was not clear what values for the template parameter `T` were allowed.

**Comment by mrzv** [comment]
@illuhad Ok, done, I think.

File: include/CL/sycl/atomic.hpp [hunk 2]
```diff
@@ -68,11 +68,33 @@ class atomic {
   /// \todo unimplemented
   HIPSYCL_KERNEL_TARGET
   void store(T operand, memory_order memoryOrder =
-      memory_order::relaxed) volatile;
+      memory_order::relaxed) volatile
+  {
+#ifdef __HIPSYCL_DEVICE_CALLABLE__
+#ifdef SYCL_DEVICE_ONLY
+    return detail::invalid_host_call_dummy_return<T>();     // really should be invalid_device_call
```
Comment (illuhad) [inline include/CL/sycl/atomic.hpp]: Better not call this on device. This function calls `assert()` which may not work on all GPUs (I think newer NVIDIA GPUs/newer CUDA versions may be able to do that, not sure about AMD).

You could instead surround the entire function with `#ifdef HIPSYCL_PLATFORM_CPU` to make it only visible when compiling for CPU.

Comment (mrzv) [reply include/CL/sycl/atomic.hpp]: That would require moving it outside the class declaration, right?

Comment (mrzv) [reply include/CL/sycl/atomic.hpp]: Would a `static_assert` be better?

Comment (mrzv) [reply include/CL/sycl/atomic.hpp]: Actually, never mind, it's not a template function, `static_assert` would get triggered every time.

Comment (illuhad) [reply include/CL/sycl/atomic.hpp]: I think an out-of-line definition outside of the class would make sense. In this case the declaration could be visible both on CPU and GPU, but the definition would be invisible on GPU, leading to a linker error.

Comment (mrzv) [reply include/CL/sycl/atomic.hpp]: How about now?

Comment (illuhad) [reply include/CL/sycl/atomic.hpp]: looks good, thank you!

**Comment by jeffhammond** [comment]
> @jeffhammond What's the right way to implement `fetch_min` and `fetch_max`? There is not an equivalent `__atomic`. The only way I can think of is via `compare_exchange`. Is there a better way?

Funny timing - I spent a bunch of time on this recently. The right implementation does CAS but only after an early-exit check for when no update is required.  This is a fundamental difference between min/max and arithmetic ops like add (all credit to @jcownie-intel for teaching me this via https://github.com/LLNL/RAJA/issues/447).

If example code helps, see https://github.com/jeffhammond/HPCInfo/blob/master/atomics/min-max/gcc-min-max.c.  I don't claim this code is perfect, so please verify correctness before using.

**Comment by mrzv** [comment]
@jeffhammond Why do you issue an extra `__atomic_load_n` inside your do-while loop? Doesn't `__atomic_compare_exchange_n` fill `old` in case of failure?

**Comment by mrzv** [comment]
@jeffhammond Also, shouldn't the `if (old > value)` check live inside the do-while loop?

**Comment by mrzv** [comment]
@jeffhammond I pushed an implementation of `fetch_min` and `fetch_max`. Can you take a look and critique?

**Comment by jeffhammond** [comment]
> @jeffhammond Why do you issue an extra __atomic_load_n inside your do-while loop? Doesn't __atomic_compare_exchange_n fill old in case of failure?
> 
Good point. It shouldn’t be there if the CAS refreshes it.

I meant it when I said my code wasn’t perfect! 😂

File: include/CL/sycl/atomic.hpp [hunk 3]
```diff
@@ -79,18 +79,34 @@ class atomic {
       memory_order::relaxed) volatile
   {
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
+#ifdef SYCL_DEVICE_ONLY
     return atomicExch(_ptr, operand);
+#else
+    return __atomic_exchange_n(_ptr, operand, __ATOMIC_RELAXED);
```
Comment (illuhad) [inline include/CL/sycl/atomic.hpp]: It's fine if we only support relaxed memory order for now, but could we maybe at least add `assert(memoryOrder == memory_order::relaxed)` on CPU? This way the user would notice that what he does is unsupported rather than the code just using a potentially different memory order.

Comment (mrzv) [reply include/CL/sycl/atomic.hpp]: Sure. I'll fix this. Currently, the only `memory_order` that exists is `relaxed`, so the assertion is vacuous, but I get the point about making it future-proof.

**Comment by illuhad** [comment]
Thank you! We also need test cases for the `atomic` class, but I can also do this later. Unless you want to, of course ;)

**Comment by mrzv** [comment]
I agree that we desperately need test cases; I've been testing this stuff in a one-off project, which is a bit of a pain. The trouble is I can't figure out your test framework. It's setup as a separate CMake project, and I haven't figured out how to get it to work. Why not just setup `make test` to run the tests?

**Comment by illuhad** [comment]
I think this probably has historic reasons.. It should be relatively easy to use though:
```
cmake -DSYCLCC=/path/to/syclcc-clang /path/to/tests
HIPSYCL_PLATFORM=cpu|cuda|rocm HIPSYCL_GPU_ARCH=<your-gpu-arch> make
```
This was written before we had our cmake integration, we should probably update it to use our current cmake integration, then we could select the target platform etc from cmake.

**Comment by mrzv** [comment]
Ok, this works for me. But let me express my enthusiastic support for updating cmake integration. (I'm afraid I'm not qualified to create a pull request for this one.)

**Comment by jeffhammond** [comment]
I will test it in isolation and then hopefully write some SYCL atomic tests, but I don't see how the atomic_min/max implementation can be correct.  I'm happy to be wrong, but see my comment above for details.

In any case, I will be happy to provide the tests and fixes once I get a few hours to work on this.

**Comment by mrzv** [comment]
@jeffhammond Can you say a little more? What do you think is wrong with it?

**Comment by jeffhammond** [comment]
I think I am just stupid.  Let me do some tests and then I'll follow-up if necessary.
